### PR TITLE
feat: add transitive reference to group export

### PR DESCRIPTION
**Pull request by Bingjiling**
Issue #, if available:

Description of changes:

* Add transitive reference to group export
* Add test for transitive reference 
* Add Device in patient compartment search param 

Checklist:

* [x] Have you successfully deployed to an AWS account with your changes?
* [x] Have you written new tests for your core changes, as applicable?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: bulkExport/schema/patientCompartmentSearchParams.4.0.1.json [hunk 1]
```diff
@@ -1 +1 @@
...
...andidate"],"EpisodeOfCare":["patient"],"ExplanationOfBenefit":["patient","payee.party"],"FamilyMemberHistory":["patient"],"Flag":["subject"],"Goal":["subject"],"Group":["member.entity"],"ImagingStudy":["subject"],"Immunization":["patient"],"ImmunizationEvaluation":["patient"],"ImmunizationRecommendation":["patient"],"Invoice":["subject","subject","recipient"],"List":["subject","source"],"MeasureReport":["subject"],"Media":["subject"],"MedicationAdministration":["subject","performer.actor","subject"],"MedicationDispense":["subject","subject","receiver"],"MedicationRequest":["subject"],"MedicationStatement":["subject"],"MolecularSequence":["patient"],"NutritionOrder":["patient"],"Observation":["subject","performer"],"Patient":["link.other"],"Person":["link.target"],"Procedure":["subject","performer.actor"],"Provenance":["target"],"QuestionnaireResponse":["subject","author"],"RelatedPerson":["patient"],"RequestGroup":["subject","action.participant"],"ResearchSubject":["individual"],"RiskAssessment":["subject"],"Schedule":["actor"],"ServiceRequest":["subject","performer"],"Specimen":["subject"],"SupplyDelivery":["patient"],"SupplyRequest":["deliverTo"],"VisionPrescription":["patient"]}
```
Comment (rsmayda) [inline bulkExport/schema/patientCompartmentSearchParams.4.0.1.json]: Could you move `Device` to be alphabetically ordered

File: integration-tests/bulkExport.test.ts [hunk 2]
```diff
@@ -81,19 +81,16 @@ describe('Bulk Export', () => {
             true,
         );
 
+        console.log(resTypToResExpectedInExport);
+        console.log(resTypToResExpectedInExport.Group.id);
+        console.log(resTypToResExpectedInExport.Organization.id);
```
Comment (rsmayda) [inline integration-tests/bulkExport.test.ts]: Do we want these log statements?

File: integration-tests/bulkExportTestHelper.ts [hunk 3]
```diff
@@ -80,7 +86,7 @@ export default class BulkExportTestHelper {
     }
 
     async getExportStatus(statusPollUrl: string, expectedSubstring = ''): Promise<any> {
-        const fiveMinuteFromNow = new Date(new Date().getTime() + this.FIVE_MINUTES_IN_MS);
+        const fiveMinuteFromNow = new Date(new Date().getTime() + this.EIGHT_MINUTES_IN_MS);
```
Comment (rsmayda) [inline integration-tests/bulkExportTestHelper.ts]:
```suggestion
        const eightMinuteFromNow = new Date(new Date().getTime() + this.EIGHT_MINUTES_IN_MS);
```

File: bulkExport/extractPatientCompartmentSearchParams.js [hunk 4]
```diff
@@ -6,7 +6,7 @@ Run the script:
 
 The compartment definition files are downloaded from the following URL and saved in folder bulkExport/schema:
 
-compartmentdefinition-patient.4.0.1.json: https://www.hl7.org/fhir/compartmentdefinition-patient.json.html
+compartmentdefinition-patient.4.0.1.json: https://www.hl7.org/fhir/compartmentdefinition-patient.json.html (Note Device is added in this file for Inferno test)
```
Comment (rsmayda) [inline bulkExport/extractPatientCompartmentSearchParams.js]:
```suggestion
compartmentdefinition-patient.4.0.1.json: https://www.hl7.org/fhir/compartmentdefinition-patient.json.html (Note Device is added to this file due to Inferno test <enter url to test>)
```

File: integration-tests/bulkExport.test.ts (hunk 2 above)
Comment (Bingjiling) [reply integration-tests/bulkExport.test.ts]: Good catch!

File: bulkExport/schema/patientCompartmentSearchParams.4.0.1.json (hunk 1 above)
Comment (Bingjiling) [reply bulkExport/schema/patientCompartmentSearchParams.4.0.1.json]: Yes! I was hoping adding it at the beginning will make it easier to spot and update. But yea it's not hard to find something ordered alphabetically. Will update this one.

File: bulkExport/extractPatientCompartmentSearchParams.js [hunk 5]
```diff
@@ -6,7 +6,7 @@ Run the script:
 
 The compartment definition files are downloaded from the following URL and saved in folder bulkExport/schema:
 
-compartmentdefinition-patient.4.0.1.json: https://www.hl7.org/fhir/compartmentdefinition-patient.json.html
+compartmentdefinition-patient.4.0.1.json: https://www.hl7.org/fhir/compartmentdefinition-patient.json.html (Note Device is added to this file due to Inferno test <enter url to test>)
```
Comment (rsmayda) [inline bulkExport/extractPatientCompartmentSearchParams.js]: sorry the "\<enter url to test\>" is suppose to link the interno test or test id

File: integration-tests/bulkExport.test.ts [hunk 6]
```diff
@@ -6,8 +6,8 @@ import BulkExportTestHelper, { ExportStatusOutput } from './bulkExportTestHelper
 import { getFhirClient } from './utils';
 import createGroupMembersBundle from './createGroupMembersBundle.json';
 
-const FIVE_MINUTES_IN_MS = 5 * 60 * 1000;
-jest.setTimeout(FIVE_MINUTES_IN_MS);
+const EIGHT_MINUTES_IN_MS = 8 * 60 * 1000;
```
Comment (rsmayda) [inline integration-tests/bulkExport.test.ts:9]: Was a longer timeout required?

File: integration-tests/bulkExportTestHelper.ts [hunk 7]
```diff
@@ -80,8 +86,8 @@ export default class BulkExportTestHelper {
     }
 
     async getExportStatus(statusPollUrl: string, expectedSubstring = ''): Promise<any> {
-        const fiveMinuteFromNow = new Date(new Date().getTime() + this.FIVE_MINUTES_IN_MS);
-        while (new Date().getTime() < fiveMinuteFromNow.getTime()) {
+        const eightMinuteFromNow = new Date(new Date().getTime() + this.EIGHT_MINUTES_IN_MS);
```
Comment (rsmayda) [inline integration-tests/bulkExportTestHelper.ts]:
```suggestion
        const cuttOffTime = new Date(new Date().getTime() + this.EIGHT_MINUTES_IN_MS);
```

Sorry thought of an even better name that isn't aligned with a specific value of the variable

File: integration-tests/bulkExport.test.ts [hunk 8]
```diff
@@ -82,18 +82,12 @@ describe('Bulk Export', () => {
...
         const groupId = resTypToResExpectedInExport.Group.id;
         const statusPollUrl = await bulkExportTestHelper.startExportJob({ exportType: 'group', groupId });
         const responseBody = await bulkExportTestHelper.getExportStatus(statusPollUrl);
 
         // CHECK
-        return bulkExportTestHelper.checkResourceInExportedFiles(
-            responseBody.output,
-            groupMembersAndPatientCompartment,
-        );
+        return bulkExportTestHelper.checkResourceInExportedFiles(responseBody.output, resTypToResExpectedInExport);
```
Comment (rsmayda) [inline integration-tests/bulkExport.test.ts:90]: Do we have an explicit test case for the transitive feature?

File: bulkExport/glueScripts/export-script.py [hunk 9]
```diff
@@ -109,42 +110,58 @@ def is_internal_reference(reference, server_url):
...
-            if isinstance(reference, dict):
-                reference = [reference]
-            elif not isinstance(reference, list):
-                return False # Inclusion criteria should point to a dict {reference: 'Patient/1234'} or a list of references
             for ref in reference:
                 if is_internal_reference(ref['reference'], server_url) and ref['reference'].split('/')[-1] in group_patient_ids:
                     return True
     return False
 
+def get_transitive_references(resource, transitive_reference_map, server_url):
```
Comment (rsmayda) [inline bulkExport/glueScripts/export-script.py:144]: Could you add a small comment explaining what a transitive reference is?

File: bulkExport/glueScripts/export-script.py [hunk 10]
```diff
@@ -157,10 +174,25 @@ def is_included_in_group_export(resource, group_member_ids, group_patient_ids, c
     print('Extract group member and patient compartment dataframe')
     filtered_group_frame = Filter.apply(frame = filtered_tenant_id_frame, f = lambda x: is_included_in_group_export(x, group_member_ids, group_patient_ids, compartment_search_params, server_url))
 
+    print('Extract forward references')
```
Comment (rsmayda) [inline bulkExport/glueScripts/export-script.py]: is `forward reference === transitive reference`?

File: bulkExport/glueScripts/export-script.py [hunk 11]
```diff
@@ -157,10 +174,25 @@ def is_included_in_group_export(resource, group_member_ids, group_patient_ids, c
...
+    transitive_reference_frame = Map.apply(frame = filtered_group_frame, f = lambda x: get_transitive_references(x, transitive_reference_params, server_url))
+    transitive_reference_frame = Filter.apply(frame = transitive_reference_frame, f = lambda x: x['_generated_transitive_refs'] is not None)
+    transitive_reference_frame = SelectFields.apply(frame = transitive_reference_frame, paths=['_generated_transitive_refs']).toDF().collect()
+
+    transitive_reference_set = set()
+    for item in transitive_reference_frame:
+        transitive_reference_set.update(reference.split('/')[-1] for reference in item['_generated_transitive_refs'])
+    print(transitive_reference_set)
```
Comment (rsmayda) [inline bulkExport/glueScripts/export-script.py]: Do we want to print this log with no context? Same for line 171 & 172

File: bulkExport/glueScripts/export-script.py [hunk 12]
```diff
@@ -157,10 +174,25 @@ def is_included_in_group_export(resource, group_member_ids, group_patient_ids, c
...
+
+    filtered_group_reference_frame = filtered_group_frame
+    if transitive_reference_set:
+        # Filter here again with transitive reference set as dynamic frame does not provide union functionality
+        filtered_group_reference_frame = Filter.apply(frame = filtered_tenant_id_frame, f = lambda x: is_included_in_group_export(x, group_member_ids, group_patient_ids, compartment_search_params, server_url, transitive_reference_ids=transitive_reference_set))
```
Comment (rsmayda) [inline bulkExport/glueScripts/export-script.py]: Anyway to just do this once? perhaps it is better just to "include" all transitive possibilities then filter them here

File: integration-tests/bulkExport.test.ts (hunk 6 above)
Comment (Bingjiling) [reply integration-tests/bulkExport.test.ts:9]: I believe so. When running the test in my dev environment, it exceeds the five minutes limit very often.

File: integration-tests/bulkExport.test.ts (hunk 8 above)
Comment (Bingjiling) [reply integration-tests/bulkExport.test.ts:90]: There's no explicit test case for transitive feature, but all group export tests covers transitive resources. The test data shared by all group export tests have 2 transitive referenced resources - Organization and Practitioner, they cover the case where the transitive reference is a dotted path and where the transitive reference point to a list.

File: bulkExport/glueScripts/export-script.py (hunk 9 above)
Comment (Bingjiling) [reply bulkExport/glueScripts/export-script.py:144]: Sure, will add it.

File: bulkExport/glueScripts/export-script.py (hunk 10 above)
Comment (Bingjiling) [reply bulkExport/glueScripts/export-script.py]: Yes! Will update the terminology here.

File: bulkExport/glueScripts/export-script.py (hunk 11 above)
Comment (Bingjiling) [reply bulkExport/glueScripts/export-script.py]: Good call! I'll add context to the print out here.

File: bulkExport/glueScripts/export-script.py (hunk 12 above)
Comment (Bingjiling) [reply bulkExport/glueScripts/export-script.py]: It's not possible to only do one filter unfortunately as this step depend on the result of the first filter. The transitive possibilities can only be deduced with the "original" export resources known, as we are looking for resources referenced from any "original" export resources, including resources from patient compartment. And the original group member can only be known from the first filter. I'll update the comment to explain it better.

File: integration-tests/bulkExport.test.ts (hunk 8 above)
Comment (rsmayda) [reply integration-tests/bulkExport.test.ts:90]: perfect thanks for confirming

File: bulkExport/glueScripts/export-script.py (hunk 12 above)
Comment (rsmayda) [reply bulkExport/glueScripts/export-script.py]: As we talked offline could you add an `if` block if someone was not to use the transitive feature at all?

**Review by rsmayda — APPROVED** [review]

**Pull request merged**